blog/tidy-finance-dummy-data/index.qmd

```diff
@@ -0,0 +1,422 @@
+  extended_types = TRUE
+)
+```
+
+Since we draw random numbers for most of the columns, we also define a seed to ensure that the generated numbers are replicable. We also initialize vectors of dates of different frequencies over 10 years that we then use to create yearly, monthly, and daily data, respectively. 
+
+```{r}
+set.seed(1234)
+
+start_date <- as.Date("2003-01-01")
+end_date <- as.Date("2022-12-31")
+
+time_series_years <- seq(year(start_date), year(end_date), 1)
+time_series_months <- seq(start_date, end_date, "1 month")
+time_series_days <- seq(start_date, end_date, "1 day")
+```
+
+## Create macro dummy data
+
+We start by creating dummy data for the macroeconomic tables. The following code does this by replacing the original data with random numbers while preserving the time structure (either daily or monthly) and column names. We iterate over each table name in the `macro_tables` vector. For each table, we fetch the data from the database and store it in a `data_original` table. The code checks the column names of `data_original` to determine if the data is structured by months or by days. Depending on the structure, it sets the appropriate time series and date column name. For each relevant column (i.e., excluding columns with dates), the code generates a command to replace its values with random numbers drawn from a uniform distribution between 0 and 1. We then use the `!!!` operator to unlist and execute the list of commands. This trick actually helps us to avoid typing the same function for each column individually. Finally, each table with dummy data is written to the new database with dummy data.
+
+Note that we do not put any meaningful structure on the macro tables because they can be freely downloaded from the original sources - check out [Accessing and Managing Financial Data](../../r/accessing-and-managing-financial-data.qmd) to replace the dummy data with the actual data.
+
+```{r}
+macro_tables <- c(
+  "cpi_monthly", "factors_ff3_daily", 
+  "factors_ff3_monthly", "factors_ff5_monthly", 
+  "factors_q_monthly", "industries_ff_monthly", 
+  "macro_predictors"
+)
+
+for (table_name in macro_tables) {
+  data_original <- tbl(tidy_finance, table_name) |>
+    collect() |> 
+    drop_na()
+  
+  if ("month" %in% names(data_original)) {
+    time_series <- time_series_months
+    date_column <- "month"
+  } else if ("date" %in% names(data_original)) {
+    time_series <- time_series_days
+    date_column <- "date"
+  }
+  
+  relevant_columns <- data_original |> 
+    select(-contains(c("month", "date"))) |> 
+    names()
+  
+  commands <- unlist(
+    map(
+      relevant_columns, 
+      ~rlang::exprs(!!..1 := runif(n()))
+    )
+  )
+  
+  data_dummy <- tibble(
+    !!sym(date_column) := time_series
+    ) |> 
+    mutate(
+      !!!commands
+    )
+  
+  dbWriteTable(
+    tidy_finance_dummy, 
+    table_name,
+    data_dummy,
+    overwrite = TRUE
+  )
+}
+```
+
+## Create stock dummy data
```

Comment:
It feels odd to not read anything about returns in this section, or am I missing something?

Reply:
It comes later at Dummy `crsp_monthly` table, but I'll add a spoiler to the first paragraph.